### ISSUE TITLE
[Accessibility]: Keyboard Support and correct aria-* mapping

### ISSUE THREAD

**Issue by rayetzki** [epam/UUI#104]
**Description**
Make sure that UUI Components are fully accessible from keyboard.

Examples:

1. Checkboxes should be focusable by pressing Tab. And by pressing Shift + Tab user can go 1 element behind.
2. Radio Buttons must be selectable by pressing Left-Right arrows.
3. Buttons must be focusable by pressing Tab or Shift Tab.
 
**Describe the solution you'd like**

- [x]  **Alert, Notification card**

    Alert top-level block must have a `role="alertdialog"` or `role="alert"`  and `aria-modal` attribute, only if alert is a modal window. By clicking `Escape`, user should be able to close the alert. Make sure that action buttons within the alert component are accessible from keyboard (Tab Sequence is preserved and buttons are clickable via `Space` or `Enter`
    ****Further reading: [https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/alertdialog.html](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/alertdialog.html)

- [x]  **Avatar Stack**

    Make sure an image container is marked via `role="group"`

- [x]  **Accordion**

    Accordions must be focusable.

    Keyboard support :

    - By pressing `Enter` or `Space` - target accordion must get expanded.
    - `Tab + Shift-Tab` - elements sequence navigation
    - `Down Arrow | Up Arrow` - Up and Down navigation between multiple accordions (if present).

    Assistive technologies use `aria-expanded` attribute to indicate, whether the accordion is expanded or not. If the accordion is disabled, the proper `aria-disabled` should also be set, since there is no native attribute. Screen readers use a combination of `aria-controls ~ id` attribute sequence to correctly map accordion header to accordion body. So, `aria-controls` with a unique id must be set on the header of the accordion, and the same value must be provided for an id attribute on the body wrapper to correctly identify containing elements. As an option, for the screen readers easier to name the accordion blocks, we can also provide a meaningful `aria-labelledby` attribute and pass a `role="region"` to the accordion body. 
    Further reading: [https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html](https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html)

- [x]  **Button (Link)**
Button must be focusable and have appropriate focus state, indicated by outline property. By pressing either Space or Enter button must get activated. If button is different from `<button></button>` , than we need to be sure, that the element is of `role="button"` or `role="link"` and proper tab order is followed (or allowed to skip it by setting `tabIndex` manually). For all buttons, that has "enable" functionality, assistive technologies require `aria-pressed` (boolean) to correctly voice the current button state. For this and other purposes `aria-label` must also have a meaningful text
Further reading: [https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html](https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html).
- [x]  **Checkbox**

    Checkbox must support keyboard press actions. Once Space is pressed, the checkbox must get checked.
    Checkboxes must have `aria-checked` and `aria-labelledby` attributes for assistive technologies to voice the label and checked status correctly.
    If the checkbox is different from `<input type="checkbox" />` , than the element must be provided with `role="checkbox"` and support proper `tab` order, using tabIndex property. Further reading: [https://www.w3.org/TR/wai-aria-practices-1.1/examples/checkbox/checkbox-1/checkbox-1.html](https://www.w3.org/TR/wai-aria-practices-1.1/examples/checkbox/checkbox-1/checkbox-1.html).

- [x]  **Control Groups (Radio, Controls, Checkboxes).**

    Must be associated via `fieldset` and optional caption.

- [x]  **Datepicker, Date Range Picker**

    Once the user focuses the date picker, the calendar must be shown. Otherwise, there is no other option to enter the calendar. Once the calendar is shown, user should automatically switch focus to current date (or minimal date if prop is set). After that, by pressing Arrow Keys, user must get navigated within the calendar. If the first day of the month is selected, by pressing `Left Arrow` user should go to the previous month. By pressing `Up Arrow` or `Down Arrow` on the first or last row, use should be able to move to the previous or next month respectivelly. By pressing `Right Arrow` user should be able to go to the next month. Within the day grid by pressing `Shift + Tab` user should be able  to go to the navigation bar (to select a year or navigate via arrow icons). Once Month-Year label is pressed the same navigation within the view must be available. By pressing `Esc` , user must leave the dialog

    Date picker dialog must have an `aria-modal="true"` attribute. Screen readers indicate a selected day via `aria-selected` attribute. Day grid must have a `role="grid"`. Controls within a date picker must have proper meaningful `aria-label` to accent the use case for each control. As an option, each datepicker section can have `aria-labelledby` property to express each section more precisely for the assistive technologies

- [x]  **Dropdown**

    To indicate that after the button gets clicked other content appears, we need to pass `aria-haspopup=true` attribute. This will tell assistive technologies, that there is content, that will appear by hovering or by clicking. Content must have an `aria-expanded` attribute set when shown

- [x]  **Dropdown Menu**

   Correct keyboard focus indication for menu items. Aria-* ```role=tree"``` for wrapper must be provided. For each individual cell within the wrapper ```role="treeitem"``` must be provided. ```aria-selected``` property must be set for the subitem if it is selected.

- [x]  **Labeled Input**

    Make sure the the element that serves as a label is either of `role="label"`, or is an actual `<label />` . The top-level element must have `aria-errormessage` that tells screen readers about possible upcoming error message. To tell assistive technologies about invalid state of the input, `aria-invalid` must be set on the element that is usually indicated with red color.

- [x]  **Main Menu**

    Make sure, that menu has a top-level element with a `role="menu"` or is a `nav` element by itself. Make sure that menu items (buttons or links) have a `role="menuitem"` 

- [x]  **Modals**

    Top-level container must have `role="dialog"` and `aria-modal=true` attributes.

    Close button must be focusable.
     
- [x]  **Multi Switch**
    
    Make sure multi switch has correct `aria-current` attribute and buttons are focusable.

- [x]  **Switch**

    Top-level element (`button`, `div` or `input[type="checkbox]` ) must have a `role="switch"` . Make sure that switch label is associated with the actual switch via `label` element. `aria-checked` must be also toggled once the element is switched.

- [x]  **Numeric Input**
- [x]  **Paginator**

    The top-level element must contain either `role="navigation"` or `nav` element. For the currently selected item, we should set `aria-current`, to make it clearer which page is currently set.

- [x]  **Picker Input**

    The input element must have an `aria-haspopup` attribute. The dialog must have an `role="dialog"`and selected elements must have `aria-selected` property. By clicking `Escape`, user must leave the picker. As an option we may do an autofocus for the search input once the dialog popups up. In addition, the picker breaks by pressing `Tab` and `Shift-Tab` which breaks the tab sequence on the page. For picker inputs with multiselection functionality, `aria-multiselectable` must be provided on the top-level of the dialog.

- [x]  **Radio Input**

    Radio buttons must support keyboard navigation. When within radio group, containing radio buttons must be switched back and forth using  ⏩ and ◀️ keys.
    Radio buttons must have `aria-checked` and `aria-labelledby` attributes for assistive technologies to voice the label and checked status correctly.
    If the radio button is different from `<input type="radio" />` , than the element must be provided with `role="radio"` and support proper `tab` order, using tabIndex property. Further reading: [https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/examples/radio/radio.html](https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/examples/radio/radio.html).

- [x]  **Rating**

    User should be able to navigate and fill in stars by pressing arrow keys. The stars must also have `aria-hidden` attribute to indicate their state for screen readers. As an addition, it would be great to discuss, whether our rating component can have a caption as an option.

- [x]  **Search Input**

    Make sure the type of input field is of `type=search` and `role="search"`.

- [x]  **Slider**

    The top-level element must have a role="slider" (`div` or `<input type=range />`). Right now, the slider doesn't have any focus indication. Outline must be shown once the element is entered from the keyboard. By using arrow keys, user should increase or decrease the slider value. For the possible values, there are `aria-valuemin`, `aria-valuemax`, `aria-valuenow` attributes that should be used to indicate current values for screen readers.

- [x]  **Range Slider**

    See Slider for a blueprint.
    
- [x]  **Slider Rating**

    See Slider for a blueprint.

- [x]  **Tab Button**

    Make sure tabs have `role=tab` , appropriate focus states with outline. By clicking `Space` or `Enter` tab button must become active. Tabs must have an `aria-selected` attribute set to indicate whether the current tab is selected or not. If the tab is within a list, than the list must have a `role="tablist"` , `aria-label` set to a caption of the tab list and an `aria-controls ~ id` combination for both tab and content, that it represents to properly map the navigation. See Accordion for example. By clicking Arrow keys, the user should navigate back-and-forth between tabs. 
    Further reading: [https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html](https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html)
    
- [x]  **Text Placeholder**
  
  Make sure text-placeholder has `aria-busy="true"` attribute 

- [x]  **Text Input**
- [x]  **Textarea**
- [x]  **Time Picker**

    See Date Picker for a blueprint.

- [x]  **Table**

    Top-level element must be a `table` or have a `role="table"` . `aria-setsize` must be present for lazy-lists to indicate how many rows may appear on the screen. `aria-rowcount`, `aria-colcount` - for row- and col- count. `aria-sort` must be set on the sort button to indicate the order in which the table is getting sorted (acsending or descending).

- [x]  **VirtualList**

    Make sure `aria-setsize` is set to the number of all possible elements that may appear in the list. 

- [x]  **Tooltip**

    To indicate that after the tooltip gets hovered, other content appears, we need to explicitly toggle `aria-hidden` attribute.
